targets/TARGET_TOSHIBA/TARGET_TMPM46B/flash_api.c

```diff
@@ -0,0 +1,163 @@
+#define FLASH_API_RAM           ((uint32_t *)__section_begin("FLASH_CODE_RAM"))
+#pragma section = "FLASH_CODE_RAM"
+#pragma section = "FLASH_CODE_ROM"
+#endif
+
+#if defined ( __ICCARM__ )    /* IAR Compiler */
+static void Copy_Routine(uint32_t * dest, uint32_t * source, uint32_t size)
+{
+    uint32_t *dest_addr, *source_addr, tmpsize;
+    uint32_t i, tmps, tmpd, mask;
+
+    dest_addr = dest;
+    source_addr = source;
+
+    tmpsize = size >> 2U;
+    for (i = 0U; i < tmpsize; i++) {    /* 32bits copy */
+        *dest_addr = *source_addr;
+        dest_addr++;
+        source_addr++;
+    }
+    if (size & 0x00000003U) {   /* if the last data size is not 0(maybe 1,2 or 3), copy the last data */
+        mask = 0xFFFFFF00U;
+        i = size & 0x00000003U;
+        tmps = *source_addr;
+        tmpd = *dest_addr;
+        while (i - 1U) {
+            mask = mask << 8U;
+            i--;
+        }
+        tmps = tmps & (~mask);
+        tmpd = tmpd & (mask);
+        *dest_addr = tmps + tmpd;       /* 32bits copy, but only change the bytes need to be changed */
+    } else {
+        /* Do nothing */
+    }
+}
+#endif
+
+int32_t flash_init(flash_t *obj)
+{
+    TSB_FC->WCLKCR  = 0x00000004U;
+    TSB_FC->PROGCR  = 0x00000001U;
+    TSB_FC->ERASECR = 0x00000007U;
+    TSB_FC->AREASEL = 0x00000000U;
+#if defined ( __ICCARM__ )    /* IAR Compiler */
+    Copy_Routine(FLASH_API_RAM, FLASH_API_ROM, SIZE_FLASH_API);
+#endif
+    return 0;
+}
+
+int32_t flash_free(flash_t *obj)
+{
+    return 0;
+}
+
+#if defined ( __ICCARM__ )    /* IAR Compiler */
+#pragma location = "FLASH_ROM"
+#endif
+int32_t flash_erase_sector(flash_t *obj, uint32_t address)
+{
+    int status = ERROR;
+    /* We need to prevent flash accesses during erase operation */
+    core_util_critical_section_enter();
+
+    if (FC_SUCCESS == FC_EraseBlock(address)) {
+        status = SUCCESS;
+    } else {
+        /* Do nothing */
+    }
+    core_util_critical_section_exit();
+    return status;
+}
+
+#if defined ( __ICCARM__ )    /* IAR Compiler */
+#pragma location = "FLASH_ROM"
+#endif
+int32_t flash_program_page(flash_t *obj, uint32_t address, const uint8_t *data, uint32_t size)
+{
+    int status = SUCCESS;
+
+    /* We need to prevent flash accesses during program operation */
+    core_util_critical_section_enter();
+    while (size > PROGRAM_WIRTE_MAX) {
+        if (FC_SUCCESS == FC_WritePage((uint32_t)address, (uint32_t *)data)) {      /* write one page every time */
+            /* Do nothing */
+        } else {
+            status = ERROR;
+            break;
+        }
+        size = size - PROGRAM_WIRTE_MAX;
+        address = address + PROGRAM_WIRTE_MAX;
+        data = data + PROGRAM_WIRTE_MAX;
+    }
+    if (FC_SUCCESS == FC_WritePage((uint32_t)address, (uint32_t *)data)) {  /* write the last data, no more than one page */
+        /* Do nothing */
+    } else {
+        status = ERROR;
+    }
+
+    core_util_critical_section_exit();
+    return status;
```

Comment:
ERROR is 1, but this API if an error is, there should be -1. Ca nyou pleas elook at HAL return types if they match?

Reply:
@0xc0170, Yes, on ERROR -1 to be returned. Added FAIL with -1 and updated.